utils/land-patterns.stanza

```diff
@@ -1071,7 +1091,156 @@ defn ipc-formula (L:Toleranced, ; the distance from edge-of-lead to edge-of-lead
+
+    defn layout-pads (dir:Dir, num-pins:Int, pitch:Double):
+      val locs = switch(dir):
+        Up: 
+          shift-locs{0.0, y-shift, _ } $
+          reverse(to-list(row-locs(num-pins, pitch)))
+        Down: 
+          shift-locs{0.0, (- y-shift), _ } $
+          row-locs(num-pins, pitch)
+        Left:  
+          shift-locs{(- x-shift), 0.0, _ } $
+          col-locs(num-pins, pitch)
+        Right: 
+          shift-locs{x-shift, 0.0, _ } $
+          reverse(to-list(col-locs(num-pins, pitch)))
+      
+      val pad-dims = 
+        if contains?([Left, Right], dir): pad-sz-ew
+        else: pad-sz-ns        
+      val idx = pin-count to (pin-count + num-pins)
+      for (i in idx, loc in locs) do:
+        pad p[i + 1]: smd-pad(pad-dims) at loc
+      pin-count = pin-count + num-pins
+
+    layout-pads(Left, num-pins-east,  pitch-east)
+    layout-pads(Down, num-pins-south, pitch-south)
+    layout-pads(Right,num-pins-west,  pitch-west)
+    layout-pads(Up,   num-pins-north, pitch-north)
+    
+    ;----------------------------------------------------------
+    ; Compute the courtyard
+    val part-sz = Dims(max-value(package-length), max-value(package-width))
+    val lp-sz   = enlarge(Dims(Zmax(ipc-ew), Zmax(ipc-ns)), 2.0 * SOLDER-MASK-REGISTRATION)
+    val cy-sz   = enlarge(max(part-sz, lp-sz), courtyard-excess(lead-fillets))
+    layer(Courtyard(Top)) = Rectangle(cy-sz)
+  
+    ;----------------------------------------------------------
+    ; Place the polarity marker
+    val line-width  = clearance(current-rules(), MinSilkscreenWidth)
+    val line-length = 3.0 * line-width - line-width
+    val dist  = SOLDER-MASK-REGISTRATION + SILK-SOLDER-MASK-CLEARANCE + line-width * 0.5 + 0.01
+    
+    val pad-loc = center(pad-1-loc)
+    val p1 = Point(
+      x(pad-loc) - 0.5 * x(pad-sz-ew),
+      y(pad-loc) + 0.5 * y(pad-sz-ew) + dist
+    )
+    val p2 = Point(
+      -0.5 * x(part-sz) - line-width * 0.5,
+       0.5 * y(part-sz) - line-width * 0.5,
+    )
+    val pol = max-distance(p1, p2)
+    layer(Silkscreen("pol", Top)) = Line(line-width, [
+      pol, pol + Point(line-length, 0.0)
+    ])
+
+    ;----------------------------------------------------------
+    ; Place the reference label
+    val text-y = 0.5 * (y(cy-sz) + MIN-SILKSCREEN-TEXT-HEIGHT)
+    ref-label(0.0, text-y)
+
+public defn make-quad-landpattern (num-pins:Int, 
+                                   pitch:Double, 
+                                   lead-span:Toleranced, 
+                                   package-size:Toleranced, 
+                                   lead-length:Toleranced, 
+                                   lead-width:Toleranced, 
+                                   lead-type:LandProtrusionType):
+  ensure-divisible!(num-pins, 4, "make-quad-landpattern(num-pins, ...)")
+  if pitch <= 0.0:  
+    fatal("make-quad-landpattern(...) failed: pitch must be greater than 0. Called with: %_." % [pitch])
```

Comment:
should not it be in the lowest level function?

Reply:
Good point, I'll update the check.